src/charm.py

```diff
@@ -160,6 +159,7 @@ def _on_ingress_changed(self, _):
         self._configure()
         self.loki_provider.update_endpoint(url=self._external_url)
         self.grafana_source_provider.update_source(source_url=self._external_url)
+        self.metrics_provider.update_scrape_job_spec(self.scrape_jobs)
```

Comment:
Have you considered
- pass `self.scrape_jobs` as a callable to MetricsEndpointProvider; **and**
- pass `self._on_ingress_changed` as a refresh event to `self.metrics_provider`?

Iiuc, in that case we wouldn't need `update_scrape_job_spec` and users won't need to remember to call it.

Reply:
Yes, none of the above examples really worked well. Take a look at this: https://github.com/canonical/prometheus-k8s-operator/pull/397#issue-1446072828